### ISSUE TITLE
WP1 DB backup is too large

### ISSUE THREAD

**Issue by rgaudin** [kiwix/operations#275]
WP1 backup on borgbase currently weights `3.34 TB`, currently increasing by 4Gb everyday.

<img width="2337" alt="Screenshot 2024-10-01 at 08 40 21" src="https://github.com/user-attachments/assets/f6311464-4375-4552-9584-70b64626af72">

It is using our default backup policy (every day for the week, one per week for two months, once per month for a year, one per year then) as visible [here](https://github.com/kiwix/operations/issues/211#issuecomment-2378988520)

[horizon](https://horizon.wikimedia.org/project/databases/4ab5abeb-7f51-444a-a484-956efe40231a/) says the DB Volume size is 75GB.

Here are a couple of insights from the DB itself

```sql
MariaDB [enwp10_prod]> SELECT table_schema "DB Name",
    -> Round(Sum(data_length + index_length) / 1024 / 1024, 1) "DB Size in MB"
    -> FROM   information_schema.tables
    -> GROUP  BY table_schema
    -> ORDER BY `DB Size in MB` DESC;
+--------------------+---------------+
| DB Name            | DB Size in MB |
+--------------------+---------------+
| enwp10_prod        |       50238.1 |
| information_schema |           0.2 |
+--------------------+---------------+
2 rows in set (0.017 sec)
```

DB size is thus ~50GB.

```sql
MariaDB [enwp10_prod]> SELECT TABLE_NAME AS "Table Name", table_rows AS "Nb Rows", ROUND( (data_length + index_length) / 1024 / 1024, 1 ) AS "Size in MB" FROM information_schema.TABLES WHERE information_schema.TABLES.table_schema = 'enwp10_prod';
+--------------------+-----------+------------+
| Table Name         | Nb Rows   | Size in MB |
+--------------------+-----------+------------+
| reviews            |     39292 |        5.0 |
| _yoyo_migration    |        29 |        0.0 |
| projects           |      2662 |        0.3 |
| cache              |      4218 |       91.9 |
| logging            | 164348413 |    42165.0 |
| workingselection   |     48246 |        2.5 |
| releases           |         0 |        0.0 |
| global_rankings    |        17 |        0.0 |
| _yoyo_version      |         0 |        0.0 |
| users              |       693 |        0.1 |
| categories         |     62362 |        7.5 |
| selections         |       264 |        0.1 |
| yoyo_lock          |         0 |        0.0 |
| passwd             |         2 |        0.0 |
| ratings            |  26508482 |     6540.9 |
| selection_data     |   3051536 |      181.9 |
| moves              |   1593591 |      540.0 |
| zim_files          |       345 |        0.1 |
| manualselectionlog |      5253 |        0.8 |
| namespacename      |     33384 |       10.5 |
| _yoyo_log          |        35 |        0.0 |
| builders           |       149 |        2.5 |
| manualselection    |      3593 |        0.3 |
| global_articles    |   7162155 |      686.0 |
| automatedselection |     44963 |        2.5 |
| custom             |         3 |        0.0 |
+--------------------+-----------+------------+
26 rows in set (0.002 sec)
```

```sql
MariaDB [enwp10_prod]> DESCRIBE logging;
+----------------------+------------------+------+-----+---------+-------+
| Field                | Type             | Null | Key | Default | Extra |
+----------------------+------------------+------+-----+---------+-------+
| l_project            | varbinary(63)    | NO   | PRI | NULL    |       |
| l_namespace          | int(10) unsigned | NO   | PRI | NULL    |       |
| l_article            | varbinary(255)   | NO   | PRI | NULL    |       |
| l_action             | varbinary(20)    | NO   | PRI | NULL    |       |
| l_timestamp          | binary(14)       | NO   | PRI | NULL    |       |
| l_old                | varbinary(63)    | YES  |     | NULL    |       |
| l_new                | varbinary(63)    | YES  |     | NULL    |       |
| l_revision_timestamp | binary(20)       | NO   |     | NULL    |       |
+----------------------+------------------+------+-----+---------+-------+
8 rows in set (0.006 sec)
```

```sql
MariaDB [enwp10_prod]> DESCRIBE ratings;
+------------------------+------------------+------+-----+---------+-------+
| Field                  | Type             | Null | Key | Default | Extra |
+------------------------+------------------+------+-----+---------+-------+
| r_project              | varbinary(63)    | NO   | PRI | NULL    |       |
| r_namespace            | int(10) unsigned | NO   | PRI | NULL    |       |
| r_article              | varbinary(255)   | NO   | PRI | NULL    |       |
| r_quality              | varbinary(63)    | YES  |     | NULL    |       |
| r_quality_timestamp    | binary(20)       | YES  |     | NULL    |       |
| r_importance           | varbinary(63)    | YES  |     | NULL    |       |
| r_importance_timestamp | binary(20)       | YES  |     | NULL    |       |
| r_score                | int(8) unsigned  | NO   |     | 0       |       |
+------------------------+------------------+------+-----+---------+-------+
8 rows in set (0.004 sec)
```

I think there are two things to sort out:

## How did we get from this to 3.34TB?

Even if we consider that all of the 26 archives in borgbase are 50GB large, this should _only_ weight 1.3TB.

- We know that our backup stores a mysql dump on borgbase, as a file
- WP1 is the only project using the _always-running_ mode of [borg-backup](https://github.com/kiwix/borg-backup/). All others using the single backup mode, triggered by a job. There might be a bug there.

## Should we backup the whole WP1 DB?

I believe WP1 keeps track of computed information that can be recomputed in the event of a crash. Maybe we should limit what we backup to what's necessary and/or not that large (I remember we already had this discussion).

**Comment by rgaudin** [comment]
FYIO horizon has a DB backup feature that I just triggered to see what comes out. I believe it's smarter than just an SQL dump but it's still in progress.

**Comment by rgaudin** [comment]
Backup is [here](https://horizon.wikimedia.org/project/database_backups/85fabd02-eefd-4468-a9bc-baf2331a6f66/). It's apparently 63.76Gb large and appears to be an [xtrabackup](https://www.percona.com/mysql/software/percona-xtrabackup) file

**Comment by benoit74** [comment]
> How did we get from this to 3.34TB?

Did you tried to get size of every archives currently in borg base? Didn't found how this can be done easily. I suspect one of them might be larger than expected for "historical" reasons.

**Comment by rgaudin** [comment]
No because I couldn't get that information easily.

I found a way to make Vortag work (had manually edit its internal DB) and got the following:

<img width="637" alt="Screenshot 2024-10-01 at 10 57 13" src="https://github.com/user-attachments/assets/d7086fe0-7c38-42d9-8823-b87657691f6b">

That's most likely the compressed size on their disk but still, there is close to no way this can lead to 3.34TB.

I have to contacted them, by email, as there is no ticket system expect a public feature request forum.

Please await for their response before digging into this.

**Comment by rgaudin** [comment]
OK here's the answer

> Hi Renaud,
>
> I have confirmed server-side that the size of this repo is indeed around 3 TB on disk. You probably need to remove some old archives and then run Compact either locally or from our web UI. Be sure to use a full access key or no data will be removed. More on this in our docs:
> <img width="445" alt="Screenshot-2024-10-01-at-13-50-43" src="https://github.com/user-attachments/assets/e226e9e8-477c-46bd-9e2a-5c61bbbf044a">
> https://docs.borgbase.com/faq/#how-can-i-prune-append-only-repositories
>
>
> Thanks,
> The BorgBase Team 
> {#HS:2722475634-2710#}

This seems coherent with old-yet-still-open https://github.com/kiwix/borg-backup/issues/57

**Comment by audiodude** [comment]
We definitely don't need the `logging` table. In fact, we probably need to remove it from the database entirely in order to make room for the article specific `page_scores` table.

Data older than 7 days is never read from that table. We have discussed in the past storing it in Redis instead: https://github.com/openzim/wp1/issues/49

**Comment by benoit74** [comment]
Should we try to manually compact another borg repository (WP1 is maybe interesting to keep as-is for now) just to confirm that we have a compacting issue which needs to be solved? Do you want me to try this?

**Comment by rgaudin** [comment]
We know from that other ticket that we do have this problem but we never tried to compact. Could we do it together?

**Comment by benoit74** [comment]
We've started compaction from the button in the UI.

Initial status:
![Image](https://github.com/user-attachments/assets/39770fcd-d6e4-4885-a3e9-69e909f6ffd5)

After compaction, this specific repo is only about 4.62GB.

Some compaction jobs are still ongoing, so no definitive figure on total size.

We've discussed we will probably plan a manual compaction every quarter (no real need to be more pushy for now, maybe every month might help at some point, tbc).

**Comment by rgaudin** [comment]
Forwarded you the discussion ; my bad, I should have copied you 🙃

**Comment by benoit74** [comment]
Thank you, email sent

**Comment by benoit74** [comment]
Compaction of `wp1db` was blocked by a lock on the repository. For some reason, it looks like we do not get an error message in such a case. I tried to remove the lock by stopping `backup` container on `mwcurator` and start again compaction but it failed, maybe I started compaction too fast after stopping backup job ... 

Anyway, now borgbase support tried again and they achieved to run compaction.

`wp1db` is now only 104.97 GB, which not optimal (it is still 10% of our total usage on borgbase currently which is now down to 1.43T) but not a concern at all anymore.

I've added a manual task in monthly routine to start this manual compaction on all our repositories. Monthly because there is no quarterly routine and because borgbase support suggested to do it weekly. Let's see if it is too cumbersome or not (should be fast when ran monthly).
